### PR TITLE
TorchDynamo: weight prepack for onednn convolution external call

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89209
* #89109
* __->__ #88988

This PR is about enabled weight prepack using the MKLDNN tensor:
1.  enable fake tensor mode for MKLDNN tensor input.
2.  make convolution fusion kernel support MKLDNN tensor input.
3. do the weight prepack at FX fusion step.

For better performance, we always use channels_last for CPU convolution path. because we test that the channels_last path can get a better performance than block input path, and also avoid the activation's layout conversion(plain to block, block to plain), currently, there only need plain to plain format conversion.

cc @VitalyFedyunin @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -3371,18 +3371,22 @@ def _prepare_convolution_fusion_create(
             [0, 0],
             groups,
         )
-        req_stride_order = get_stride_order(output.stride())
+        output_size = output.size()
+        stride_order = [0] + list(reversed(range(1, 2 + 1)))
```
Comment (jgong5) [inline torch/_inductor/ir.py:3375]: nit: is `2` the kernel size?

File: torch/_inductor/overrides.py [hunk 2]
```diff
@@ -327,31 +363,36 @@ def forward(self, input, other):
...
     assert not (conv.training), "Fusion only for eval!"
     return ConvBinaryInplace2d(
         conv,
         binary_op_name,
+        input_size,
     )
 
 
-def fused_binary_unary_eval(conv_binary: nn.Module, unary: nn.Module):
+def fused_binary_unary_eval(conv_binary: nn.Module, unary: nn.Module, input_size: list):
```
Comment (jgong5) [inline torch/_inductor/overrides.py:395]: nit: `fused_conv_binary_unary_eval` instead?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/mkldnn/Utils.cpp [hunk 3]
```diff
@@ -38,13 +38,15 @@ void check_mkldnn_binary_fusion_inputs(
...
-      ") should be the same");
+  if (!weight.is_mkldnn()) {
+    TORCH_CHECK(
+        input.options().type_equal(weight.options()),
+        "Input type (",
+        input.toString(),
+        ") and weight type (",
+        weight.toString(),
+        ") should be the same");
+  }
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply aten/src/ATen/native/mkldnn/Utils.cpp:49]: Yes, we can add dtype check.

File: torch/_inductor/ir.py (hunk 1 above)
Comment (XiaobingSuper) [reply torch/_inductor/ir.py]: Yes, using stride's size now.

File: torch/_inductor/overrides.py (hunk 2 above)
Comment (XiaobingSuper) [reply torch/_inductor/overrides.py]: Chaneged.

**Review by jgong5 — APPROVED** [review]

**Comment by desertfire** [comment]
> For better performance, we always use channels_last for CPU convolution path. because we test that the channels_last path can get a better performance than block input path

Is this conclusion always true? Does the better performance depend on conditions like input shapes?

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -3374,18 +3374,20 @@ def _prepare_convolution_fusion_create(
             [0, 0],
             groups,
         )
-        req_stride_order = get_stride_order(output.stride())
+        output_size = output.size()
+        req_stride_order = [0] + list(reversed(range(1, len(stride) + 1)))
+        req_stride_order = [len(req_stride_order)] + req_stride_order
+        ouput_stride = make_channels_last_strides_for(output_size)
```
Comment (desertfire) [inline torch/_inductor/ir.py:3380]: Nit: typo, `output_stride`

**Comment by EikanWang** [comment]
> Does the better performance depend on conditions like input sha

Compared with PyTorch's current solution, it might not be true for some particular shapes but it should always be true from the E2E perspective. Channels-first Conv uses blocked format for computation and converts blocked format tensor to channels-first.

**Comment by jgong5** [comment]
> > Does the better performance depend on conditions like input sha
> 
> Compared with PyTorch's current solution, it might not be true for some particular shapes but it should always be true from the E2E perspective. Channels-first Conv uses blocked format for computation and converts blocked format tensor to channels-first.

Right. For individual convs, there are some special shapes for which THNN implementation is used (https://github.com/jgong5/pytorch/blob/407d93bd7911f54460831a68cfec9fe5a09c733c/aten/src/ATen/native/Convolution.cpp#L509) that works directly with channels-first format and is faster than oneDNN conv with channel-first (oneDNN requires an extra layout conversion to blocked). But as @EikanWang pointed out, channels last benefits end-to-end workload in general since 1) little chance for all those conv shapes to be special shapes; 2) the format can propagate through the model without conversions in the middle of the model.

**Review by jansel — APPROVED** [review]

File: torch/_inductor/ir.py (hunk 4 above)
Comment (XiaobingSuper) [reply torch/_inductor/ir.py:3380]: Yes, changed.

**Pull request closed without merging**